### PR TITLE
fix(simulation): set_joint_positions refuses a pose outside the joint range

### PR DESCRIPTION
**Pull request by cagataycali**
so100 base joint driven through its limit, before and after

**What**
`set_joint_positions` refuses a pose outside a limited joint's range, naming joint, value and range; nothing is written. A joint with no `range` still takes any finite value.

**Why**
`mj_forward` does not clamp `qpos`. On the bundled `so100`, `[99, 0, 0, 0, 0, 0]` on the `[-1.92, 1.92]` base joint returned success (`Set 6/6 joint positions, FK updated`); 100 steps then drove it to -9.44 rad at 23.8 rad/s, peaking at 2892 rad/s. Refuse-not-clamp, as #3425 does for the G1.

**Tests**
6 new cells, 4 fail pre-fix: limits accepted, one bad value refuses the whole write, an unlimited joint unaffected (no limited ball/free joint exists across the 63-robot catalog). 3 pre-existing tests that teleported past a limit now use range midpoints. Full `tests/` on aarch64: 50,841 passed / 303 skipped. Merged on main, `tests/simulation` gives 15,012 vs 15,006 passed (+6, the new cells); the lone failure, `test_policy_runner_video_writer_cleanup`, reproduces identically on unmodified main.

---

## Review changelog

| Round | Concern | Fix commit | Pin |
|---|---|---|---|
| R1 | Changelog fragment named with the `0000` placeholder instead of the PR number; the `>= 9000` placeholder guard has no low-side twin, so CI was green while the release-note pointer would have been lost at `--apply`. One `git mv`, content unchanged. | `27ba70a4` | `tests/test_changelog_fragments.py` 28 passed on the renamed tree |

### CODE REVIEW

File: changelog.d/3429-set-joint-positions-refuse-out-of-range.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+### Fixed: `set_joint_positions` refuses a pose outside a limited joint's range
```
Comment (yinsong1986) [inline changelog.d/3429-set-joint-positions-refuse-out-of-range.md:1]: [MUST FIX] The fragment is named with the placeholder number `0000` rather than this PR's number. The naming contract (`changelog.d/README.md`, AGENTS.md > PR Workflow step 3) is `<number>-<slug>.md` where `<number>` is the PR or issue number - rename to `3429-set-joint-positions-refuse-out-of-range.md`.

Why this can't be deferred: at release time `assemble_changelog.py --apply` deletes the fragment as it folds it into `CHANGELOG.md`, so the number is the *only* pointer from the release-note entry back to this change, and once merged under `0000` it is unrecoverable - the exact irreversibility `tests/test_changelog_fragments.py::test_no_fragment_uses_a_reserved_placeholder_number` exists to refuse ("renaming it before merge is easy to forget: nothing downstream fails"). That guard only catches placeholders `>= 9000`, which is why CI is green here; `0000` escapes on the low side and additionally sorts below every real entry in the assembled section (assembly orders by descending number). One `git mv` resolves it.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

Adds a range guard to `set_joint_positions`: a pose outside a limited joint's `range` is now refused before any `qpos` write, with an error naming each offending joint, value and range, while unlimited joints keep accepting any finite value. This closes the `mj_forward`-does-not-clamp hole where an out-of-range write reported success and the next step drove the joint back through its limit at solver-produced velocity. Refuse-not-clamp matches the house rule and the #3425 precedent for the G1.

## What's good

- The refusal returns a structured error dict (never raises), keeps the all-or-nothing write contract, runs before the `qpos` write, and the write itself stays under `self._lock`; the pre-lock model reads (`jnt_limited`, `jnt_range`) match the function's existing pre-lock resolution pattern.
- New test file covers far/just-below/just-above refusals, whole-write atomicity on one bad value, both limits accepted, and the unlimited-joint over-reach control - with an inline MJCF fixture, no network dependency.
- The parity test (`send_action` vs `set_joint_positions`) is updated to state exactly where the two writers' domains legitimately diverge (ctrl clamps to `ctrlrange`, qpos has nothing to clamp to), and the pre-existing tests that teleported past limits now derive range midpoints from the model rather than hardcoding poses.
- Error text is plain ASCII and names an actionable remedy.

File: changelog.d/3429-set-joint-positions-refuse-out-of-range.md (hunk 1 above)
Comment (cagataycali) [reply changelog.d/3429-set-joint-positions-refuse-out-of-range.md:1]: Renamed in `27ba70a4`: `git mv changelog.d/0000-set-joint-positions-refuse-out-of-range.md changelog.d/3429-set-joint-positions-refuse-out-of-range.md`, content unchanged. `tests/test_changelog_fragments.py` 28 passed on the renamed tree. Agreed on the irreversibility point - the number is the only pointer left after `--apply` deletes the fragment, and the `>= 9000` guard has no low-side twin, which is why `0000` was green.

**Review by yinsong1986 — APPROVED** [review]
## Summary

Closes the `mj_forward`-does-not-clamp hole in `set_joint_positions`: a pose outside a limited joint's `range` is now refused before any `qpos` write, with an error naming each offending joint, value and range, while a joint with no `range` keeps accepting any finite value. The guard sits after the existing finiteness/name-resolution checks and before the lock, reads only static model fields (`jnt_limited` / `jnt_range`), and returns a structured error dict per the tool contract, so a refusal leaves `qpos` byte-identical - the new test pins exactly that. The refuse-not-clamp shape matches the house rule and the #3425 precedent for the G1, and no production caller in the MuJoCo path writes through `set_joint_positions`, so the new refusal cannot silently break a reset or home helper. The prior review's must-fix (fragment named `0000` instead of the PR number) is resolved in `27ba70a4` as a pure rename (R100, content unchanged).

## What's good

- The new test file carries the measured failure in its docstring (99 rad on the so100 base joint, -9.4 rad at 23.8 rad/s after 100 steps) and includes the over-reach control: an unlimited joint still accepts any finite value.
- All-or-nothing semantics pinned: one bad value refuses the whole write, with a `qpos`-unchanged assertion, and boundary values exactly at the limits are pinned as accepted.
- The parity test with `send_action` documents *why* the two writers legitimately diverge outside the range (ctrl clamps to `ctrlrange`, qpos has nothing to clamp to) instead of silently loosening the parity claim, and its `[-1.5, 1.5]` window matches the inline fixture XML's joint range.
- The three pre-existing tests that teleported past a limit now derive range midpoints from the model rather than hardcoding per-robot constants, so a registry range change cannot silently re-break them.
- Error text is plain ASCII and ends with an actionable remedy, per AGENTS.md > Unicode & String Hygiene and the error-dict contract.

## Verification suggestions

Beyond CI: `pytest tests/simulation/mujoco/test_set_joint_positions_refuses_out_of_range.py tests/simulation/mujoco/test_send_action_finite_values.py tests/simulation/mujoco/test_input_validation.py tests/simulation/mujoco/test_inject_failure_cleanup.py tests/test_changelog_fragments.py -q` on the head tree - reproduced green here (37 + 139 passed, changelog grader 28 passed on the renamed fragment).
